httomo/wrappers_class.py

```diff
@@ -0,0 +1,312 @@
+from typing import Dict
+import numpy as np
+import inspect
+from mpi4py import MPI
+from mpi4py.MPI import Comm
+from inspect import signature
+from httomo.utils import print_once
+
+
+gpu_enabled = False
+try:
+    import cupy as xp
+
+    try:
+        xp.cuda.Device(0).compute_capability
+        gpu_enabled = True  # CuPy is installed and GPU is available
+    except xp.cuda.runtime.CUDARuntimeError:
+        import numpy as xp
+
+        print("CuPy is installed but GPU device inaccessible")
+except ImportError:
+    import numpy as xp
+
+    print("CuPy is not installed")
+
+
+class BaseWrapper:
+    """A parent class for all wrappers in httomo that use external modules."""
+
+    def __init__(
+        self, module_name: str, function_name: str, method_name: str, comm: Comm
+    ):
+        self.comm = comm
+        local_comm = comm.Split_type(MPI.COMM_TYPE_SHARED)
+        if gpu_enabled:
+            self.num_GPUs = xp.cuda.runtime.getDeviceCount()
+            self.gpu_id = local_comm.rank % self.num_GPUs
+            xp._default_memory_pool.free_all_blocks()
+
+    def _transfer_data(self, *args) -> tuple:
+        """transfers data between the host and device
+        Returns:
+            tuple: converted datasets
+        """
+        ret = tuple()
+        for datasets in args:
+            if gpu_enabled:
+                xp.cuda.Device(self.gpu_id).use()
+                if self.cupyrun:
+                    # the method accepts CuPy arrays for the GPU processing
+                    # move the data to the current device
+                    ret = ret + (xp.asarray(datasets),)
+                else:
+                    # the method doesn't accept CuPy arrays
+                    ret = ret + (xp.asnumpy(datasets),)
```

Comment:
and what if `gpu_enabled` is `False`?

Reply:
but you're returning an empty tuple in that case...

Reply:
oops!